src/ameba/rule/typing/macro_call_argument_type_restriction.cr

```diff
@@ -0,0 +1,92 @@
+module Ameba::Rule::Typing
+  # A rule that enforces variable arguments to specific macros have a type restriction.
+  # By default these macros are: `(class_)getter/setter/property(?/!)` and `record`.
+  #
+  # For example, these are considered invalid:
+  #
+  # ```
+  # class Greeter
+  #   getter name
+  #   getter age = 0.days
+  # end
+  #
+  # record Task,
+  #   cmd = "",
+  #   args = %w[]
+  # ```
+  #
+  # And these are considered valid:
+  #
+  # ```
+  # class Greeter
+  #   getter name : String?
+  #   getter age : Time::Span = 0.days
+  # end
+  #
+  # record Task,
+  #   cmd : String = "",
+  #   args : Array(String) = %w[]
+  # ```
+  #
+  # The `DefaultValue` configuration option controls whether this rule applies to
+  # call arguments that have a default value.
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/MacroCallArgumentTypeRestriction:
+  #   Enabled: false
+  #   DefaultValue: true
+  #   MacroNames:
+  #    - getter
+  #    - getter?
+  #    - getter!
+  #    - class_getter
+  #    - class_getter?
+  #    - class_getter!
+  #    - setter
+  #    - setter?
+  #    - setter!
+  #    - class_setter
+  #    - class_setter?
+  #    - class_setter!
+  #    - property
+  #    - property?
+  #    - property!
+  #    - class_property
+  #    - class_property?
+  #    - class_property!
+  #    - record
+  # ```
+  class MacroCallArgumentTypeRestriction < Base
+    properties do
+      since_version "1.7.0"
+      description "Recommends that call arguments to certain macros have type restrictions"
+      enabled false
+      default_value true
+      macro_names %w[
+        getter getter? getter! class_getter class_getter? class_getter!
+        setter setter? setter! class_setter class_setter? class_setter!
+        property property? property! class_property class_property? class_property!
+        record
+      ]
+    end
+
+    MSG = "Argument should have a type restriction"
+
+    def test(source, node : Crystal::Call)
+      return unless node.name.in?(macro_names)
+
+      node.args.each do |arg|
+        case arg
+        when Crystal::Assign
+          next unless default_value?
+
+          issue_for arg.target, MSG, prefer_name_location: true
+        when Crystal::Var, Crystal::Call
```

Comment:
In what circumstances the `Crystal::Call` is being passed into these macros?
And even if it is, how can it have a type restriction?

Reply:
In `getter name`, `name` is parsed as a call node.

```crystal
icr:13> macro my_macro(a)
icr:14>   {% puts a.is_a?(Call) %}
icr:15> end
 => nil
icr:16> my_macro name
true
 => nil
```

Reply:
I see, thx.

Reply:
Btw, `StringLiteral` and `SymbolLiteral` are supported as well by the getter/setter/property macros.

Reply:
Good point, I'll add specs for those